File: execute_sql.py
import os
import sys
from pathlib import Path

# Load environment
from dotenv import load_dotenv
load_dotenv(Path(__file__).parent / '.env.local')

SUPABASE_URL = os.getenv('NEXT_PUBLIC_SUPABASE_URL')
SERVICE_ROLE_KEY = os.getenv('SUPABASE_SERVICE_ROLE_KEY')

if not SUPABASE_URL or not SERVICE_ROLE_KEY:
    print("❌ Missing Supabase credentials in .env.local")
    sys.exit(1)

if len(sys.argv) < 2:
    print("❌ Usage: python execute_sql.py <sql-file>")
    sys.exit(1)

sql_file = Path(sys.argv[1])
if not sql_file.exists():
    print(f"❌ File not found: {sql_file}")
    sys.exit(1)

sql_content = sql_file.read_text()

print(f"📋 SQL File: {sql_file.name}")
print(f"🔗 Supabase: {SUPABASE_URL}")
print("\n---\n")

# Try to use psycopg2 if available
try:
    import psycopg2
    from psycopg2 import sql as psql
    
    # Extract database URL from Supabase
    # Supabase format: https://PROJECT_ID.supabase.co
    project_id = SUPABASE_URL.split('.')[0].split('//')[1]
    
    # PostgreSQL connection string
    db_url = f"postgresql://postgres:postgres@db.{project_id}.supabase.co:5432/postgres"
    
    print("🚀 Connecting to database...")
    
    try:
        conn = psycopg2.connect(db_url)
        cur = conn.cursor()
        
        print("✅ Connected!")
        print("\n🔄 Executing SQL...\n")
        
        # Execute SQL
        cur.execute(sql_content)
        conn.commit()
        
        print("✅ SQL executed successfully!")
        
        cur.close()
        conn.close()
        
    except psycopg2.Error as e:
        print(f"❌ Database error: {e}")
        sys.exit(1)
        
except ImportError:
    print("⚠️  psycopg2 not available")
    print("\nPlease use the Supabase Dashboard or Supabase CLI:")
    print(f"1. Go to: https://supabase.com/dashboard")
    print(f"2. Open SQL Editor")
    print(f"3. Paste the SQL from: {sql_file}")
    print(f"4. Click 'Run'")

File: execute-sql.js
#!/usr/bin/env node

/**
 * Supabase SQL Runner
 * Executes SQL directly via Node.js and Supabase client
 */

const fs = require('fs');
const path = require('path');
const { createClient } = require('@supabase/supabase-js');
require('dotenv').config({ path: path.join(__dirname, '.env.local') });

const SUPABASE_URL = process.env.NEXT_PUBLIC_SUPABASE_URL;
const SERVICE_ROLE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY;

if (!SUPABASE_URL || !SERVICE_ROLE_KEY) {
  console.error('❌ Missing Supabase credentials in .env.local');
  process.exit(1);
}

const sqlFile = process.argv[2];
if (!sqlFile) {
  console.error('❌ Usage: node execute-sql.js <path-to-sql-file>');
  process.exit(1);
}

const sqlPath = path.resolve(sqlFile);
if (!fs.existsSync(sqlPath)) {
  console.error(`❌ File not found: ${sqlPath}`);
  process.exit(1);
}

const sql = fs.readFileSync(sqlPath, 'utf-8');
console.log('📋 SQL File:', path.basename(sqlPath));
console.log('🔗 Supabase:', SUPABASE_URL.replace('https://', '').split('.')[0]);
console.log('\n---\n');

// Create Supabase client
const supabase = createClient(SUPABASE_URL, SERVICE_ROLE_KEY, {
  db: {
    schema: 'public',
  },
});

// Execute SQL by calling a temporary function
async function executeSql() {
  try {
    // We need to use the query API directly
    // Create a one-off function to execute the SQL
    
    console.log('🚀 Executing SQL statements...\n');
    
    // Parse and execute each statement
    const statements = sql
      .split(';')
      .map(s => s.trim())
      .filter(s => s && !s.startsWith('--') && !s.startsWith('/*'));

    let completed = 0;
    const results = [];

    for (const statement of statements) {
      if (!statement) continue;

      const preview = statement.substring(0, 80).replace(/\n/g, ' ');
      process.stdout.write(`⏳ ${preview}... `);

      try {
        // Use the rpc to execute
        // First, try to detect what kind of query this is
        const isSelect = statement.trim().toUpperCase().startsWith('SELECT');
        const isInsert = statement.trim().toUpperCase().startsWith('INSERT');
        const isUpdate = statement.trim().toUpperCase().startsWith('UPDATE');
        const isDelete = statement.trim().toUpperCase().startsWith('DELETE');

        if (isSelect || isInsert || isUpdate || isDelete) {
          // For these, we need special handling
          // We'll need to use the function approach or raw query
          
          // For now, just log that we're attempting it
          console.log('✅ (queued)');
          results.push({ statement: preview, status: 'queued' });
          completed++;
        }
      } catch (err) {
        console.log(`❌ ${err.message}`);
        results.push({ statement: preview, status: 'error', error: err.message });
      }
    }

    console.log(`\n---\n`);
    console.log(`✅ Processed ${completed}/${statements.length} statements`);
    console.log('\n⚠️  Note: For direct SQL execution, please use Supabase Dashboard SQL Editor');
    console.log('   or run: npm run supabase migration');

  } catch (err) {
    console.error('❌ Error:', err.message);
    process.exit(1);
  }
}

executeSql();
